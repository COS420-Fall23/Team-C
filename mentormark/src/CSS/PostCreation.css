.page-center {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
  padding: 20px;
  background-color: #edede8;
}

.creation-container {
  display: flex;
  flex-direction: column;
  align-content: center;
  padding: 40px;
  max-width: 500px;
  background-color: #ffffff;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
  border-radius: 10px;
}

.create-title {
  color: rgb(97, 97, 220);
  font-size: 40px;
  margin-bottom: 70px;
  align-self: center;
}

.create-form {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.create-label {
  color: rgb(134, 134, 216);
  margin-bottom: 11px;
  font-size: 20px;
  font-weight: bold;
  text-align: left;
}

.create-input[type="text"],
.create-textarea,
.create-input[type="file"] {
  border: 1px solid #ccc;
  border-radius: 5px;
  padding: 10px;
  margin-bottom: 15px;
  width: 100%;
}

.create-textarea {
  resize: vertical;
  min-height: 100px;
}

.create-button {
  background-color: rgb(202, 131, 238);
  color: white;
  padding: 10px 15px;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  font-weight: bold;
}

.create-button:hover {
  background-color: rgb(134, 134, 216);
}

.create-home {
  background-color: rgb(150, 12, 214);
  color: white;
  padding: 10px 15px;
  border: none;
  border-radius: 5px;
  max-width: 100px;
  cursor: pointer;
  font-weight: bold;
  align-self: center;
}

.create-home:hover {
  background-color: rgb(134, 134, 216);
}

