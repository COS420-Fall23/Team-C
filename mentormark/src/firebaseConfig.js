// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";

import { getAuth, updateProfile, updateEmail, updatePassword, deleteUser, setPersistence, browserSessionPersistence } from 'firebase/auth';
import { doc, getDoc, getFirestore, updateDoc, deleteDoc, collection } from 'firebase/firestore';
import { getStorage, ref, uploadBytesResumable, getDownloadURL } from "firebase/storage";
// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: "AIzaSyAKpuL6t9ihiA-6gsLwCZAN-TvChkJSAfo",
  authDomain: "mentormark-ace5a.firebaseapp.com",
  projectId: "mentormark-ace5a",
  storageBucket: "mentormark-ace5a.appspot.com",
  messagingSenderId: "791836430319",
  appId: "1:791836430319:web:fe5c9b8f57a79a4968f01a",
  measurementId: "G-1KFNCC5NBQ"
};

export async function getUserProfilePictureFromFirestore(userId) {
  try {
    const userDoc = doc(collection(db, 'users'), userId);
    const docSnap = await getDoc(userDoc);

    if (docSnap.exists()) {
      const userData = docSnap.data();
      return userData.profilePicture; // Assuming profilePicture is a field in the user document
    } else {
      // Handle if user document doesn't exist
      console.log('User document does not exist');
      return ''; // Return an empty string or default URL if needed
    }
  } catch (error) {
    // Handle error if fetching profile picture fails
    console.error('Error fetching user profile picture:', error);
    return ''; // Return an empty string or default URL if needed
  }
}


// Initialize Firebase
const app = initializeApp(firebaseConfig);

export const auth = getAuth(app);
export const db = getFirestore(app);
export const storage = getStorage(app);
setPersistence(auth, browserSessionPersistence);
export { updateProfile, updateEmail, updatePassword, deleteUser, doc, getDoc, updateDoc, deleteDoc, ref, uploadBytesResumable, getDownloadURL, collection};

