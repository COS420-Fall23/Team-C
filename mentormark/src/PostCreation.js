import React, { useState, useEffect } from "react";
import "./CSS/PostCreation.css";
import { useNavigate } from "react-router-dom";
import { storage } from "./firebaseConfig.js";
import { ref, uploadBytesResumable, getDownloadURL } from "firebase/storage";
import { collection, addDoc } from "firebase/firestore";
import { getAuth, onAuthStateChanged } from "firebase/auth";
import { db } from "./firebaseConfig.js";
import "./components/Comment.js"



function PostCreation() {
  const [formData, setFormData] = useState({
    title: '',
    content: '',
    community: '',
    userID: '',
    file: null,
  });
  const [percent, setPercent] = useState(0);

  const navigate = useNavigate();
  

  const handleInputChange = (e) => {
    const { name, value, files } = e.target;
    setFormData({
      ...formData,
      [name]: name === 'file' ? files[0] : value,
    });
  };

  const auth = getAuth();
  const [user, setUser] = useState(null);

  // Listen for changes in authentication state
  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      setUser(user);
    });

    // Cleanup function
    return () => unsubscribe();
  }, [auth]);

  const createPost = async () => {
    const { title, content, community, file } = formData;
  
    if (title && content && community) {
      const newPost = {
        title,
        content,
        community,
        userID: user.uid,
        timestamp: new Date().toLocaleString(),
      };
      console.log(user.uid);
  
      try {
        if (file) {
          const storageRef = ref(storage, `/files/${file.name}`);
          const uploadTask = uploadBytesResumable(storageRef, file);
  
          uploadTask.on(
            "state_changed",
            (snapshot) => {
              const percent = Math.round(
                (snapshot.bytesTransferred / snapshot.totalBytes) * 100
              );
              setPercent(percent);
            },
            async (err) => console.log(err),
            async () => {
              const url = await getDownloadURL(uploadTask.snapshot.ref);
              newPost.file = url;
  
              const docRef = await addDoc(collection(db, 'posts'), newPost);
              console.log('Document written with ID: ', docRef.id);
              navigate('/mainpage'); // Redirect back to the main page
            }
          );
        } else {
          const docRef = await addDoc(collection(db, 'posts'), newPost);
          console.log('Document written with ID: ', docRef.id);
          navigate('/mainpage'); // Redirect back to the main page
        }
      } catch (e) {
        console.error('Error adding document: ', e);
      }
    } else {
      alert('Please provide the title, content, and community for the post.');
    }
  };
  
  return (
    <div className="page-center">
      <div className="creation-container">
        <button className="create-home" onClick={() => navigate('/mainpage')}>Home</button>
        <h1 className="create-title">Create Post</h1>
        <form className="create-form">
          <label className="create-label" htmlFor="post-title">Title:</label>
          <input
            className="create-input"
            type="text"
            id="post-title"
            name="title"
            value={formData.title}
            onChange={handleInputChange}
            required
          />
          <br />
          <label className="create-label" htmlFor="post-content">Content:</label>
          <textarea
            className="create-textarea"
            id="post-content"
            name="content"
            value={formData.content}
            onChange={handleInputChange}
            required
          ></textarea>
          <br />
          <label className="create-label" htmlFor="post-community">Community:</label>
          <select 
            className="create-select"
            name="community"
            id="post-community"
            value={formData.community} 
            onChange={handleInputChange} 
            required>
            <option value="null">--</option>
            <option value="Computer Science">Computer Science</option>
            <option value="New Media Design">New Media Design</option>
          </select>
          <label className="create-label" htmlFor="post-file">Upload File:</label>
          <input
            className="create-input"
            type="file"
            id="post-file"
            name="file"
            onChange={handleInputChange}
          />
          <br />
          <button className="create-button" type="button" onClick={createPost}>
            Create Post
          </button>
        </form>
      </div>
    </div>
  );
};

export default PostCreation;


// Post must be created in a community