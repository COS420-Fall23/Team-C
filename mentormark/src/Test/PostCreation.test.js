import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { useNavigate } from 'react-router-dom';
import { storage, db } from '../firebaseConfig.js';
import PostCreation from '../PostCreation';

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

jest.mock('../firebaseConfig.js', () => ({
  storage: {
    ref: jest.fn((storage, path) => ({
      uploadBytesResumable: jest.fn().mockReturnValue({
        on: jest.fn((event, progress, error, success) => {
          if (event === 'state_changed') {
            progress({ bytesTransferred: 100, totalBytes: 200 });
            success();
          }
        }),
        snapshot: {
          ref: {
            getDownloadURL: jest.fn().mockResolvedValue('mock-url'),
          },
        },
      }),
    })),
  },
  db: {
    collection: jest.fn(() => ({
      addDoc: jest.fn().mockResolvedValue({ id: 'mock-id' }),
    })),
  },
}));

describe('<PostCreation />', () => {
  test('renders without crashing', () => {
    render(<PostCreation />);
  });

  test('renders form with title, content and file input', () => {
    render(<PostCreation />);
    expect(screen.getByLabelText('Title:')).toBeInTheDocument();
    expect(screen.getByLabelText('Content:')).toBeInTheDocument();
    expect(screen.getByLabelText('Upload File:')).toBeInTheDocument();
  });

  test('calls handleInputChange when title input changes', () => {
    render(<PostCreation />);
    fireEvent.change(screen.getByLabelText('Title:'), { target: { value: 'Test Title' } });
    expect(screen.getByLabelText('Title:').value).toBe('Test Title');
  });

  test('calls handleInputChange when content input changes', () => {
    render(<PostCreation />);
    fireEvent.change(screen.getByLabelText('Content:'), { target: { value: 'Test Content' } });
    expect(screen.getByLabelText('Content:').value).toBe('Test Content');
  });

  test('calls createPost when Create Post button is clicked', async () => {
    render(<PostCreation />);
    fireEvent.change(screen.getByLabelText('Title:'), { target: { value: 'Test Title' } });
    fireEvent.change(screen.getByLabelText('Content:'), { target: { value: 'Test Content' } });
    fireEvent.change(screen.getByLabelText('Upload File:'), { target: { files: [new File(['file'], 'test.png', { type: 'image/png' })] } });
    const button = screen.getByRole('button', { name: /Create Post/i });
    fireEvent.click(button);
    await waitFor(() => expect(useNavigate).toHaveBeenCalled());
  });

  test('shows alert when Create Post button is clicked with incomplete form', () => {
    window.alert = jest.fn();
    render(<PostCreation />);
    const button = screen.getByRole('button', { name: /Create Post/i });
    fireEvent.click(button);
    expect(window.alert).toHaveBeenCalled();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<PostCreation />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders form with a submit button', () => {
    render(<PostCreation />);
    const button = screen.getByRole('button', { name: /Create Post/i });
    expect(button).toBeInTheDocument();
  });

  test('title input is initially empty', () => {
    render(<PostCreation />);
    expect(screen.getByLabelText('Title:').value).toBe('');
  });

  test('content textarea is initially empty', () => {
    render(<PostCreation />);
    expect(screen.getByLabelText('Content:').value).toBe('');
  });
});