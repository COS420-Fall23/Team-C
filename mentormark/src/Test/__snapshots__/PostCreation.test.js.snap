// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostCreation /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="page-center"
  >
    <div
      class="creation-container"
    >
      <button
        class="create-home"
      >
        Home
      </button>
      <h1
        class="create-title"
      >
        Create Post
      </h1>
      <form
        class="create-form"
      >
        <label
          class="create-label"
          for="post-title"
        >
          Title:
        </label>
        <input
          class="create-input"
          id="post-title"
          name="title"
          required=""
          type="text"
          value=""
        />
        <br />
        <label
          class="create-label"
          for="post-content"
        >
          Content:
        </label>
        <textarea
          class="create-textarea"
          id="post-content"
          name="content"
          required=""
        />
        <br />
        <label
          class="create-label"
          for="post-file"
        >
          Upload File:
        </label>
        <input
          class="create-input"
          id="post-file"
          name="file"
          type="file"
        />
        <br />
        <button
          class="create-button"
          type="button"
        >
          Create Post
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
